### ISSUE TITLE
Illegal access exception with macros (JRuby 9.1.15.0 & Java 9)

### ISSUE THREAD

**Issue by jmini** [asciidoctor/asciidoctorj#624]
When you run a project with a simple macro (I have taken the java-extension-example project from the maven examples) with JRuby 9.1.15.0 (latest) and Java 9 you get an Exception like this:

> [ERROR] Failed to execute goal org.asciidoctor:asciidoctor-maven-plugin:1.5.6:process-asciidoc (asciidoctor-with-extension-example) on project asciidoctor-with-extension-example: Execution asciidoctor-with-extension-example of goal org.asciidoctor:asciidoctor-maven-plugin:1.5.6:process-asciidoc failed: org.jruby.exceptions.RaiseException: (TypeError) illegal access on 'process': class org.jruby.javasupport.JavaMethod cannot access a member of class org.asciidoctor.extension.MacroProcessor with modifiers "protected abstract" -> [Help 1]

I did not investigate it yet, but I posted the issue here because I think the problem comes from AsciidoctorJ and the newer versions of JRuby and the JVM and not from the specific maven example.

The issue does not appear with:

* Java 8 & JRuby 9.1.15.0
* Java 8 & JRuby 9.0.5.0
* Java 9 & JRuby 9.0.5.0

Tested on macOS.

**Comment by abelsromero** [comment]
In case it helps, the issue appears with jRuby 9.1.14.0 which is the first marked as supporting Java9 (http://jruby.org/news).
Previous versions that work fine for me: Maven 3.5.2, Java 9.0.4, Windows7.

```
Caused by: org.jruby.exceptions.RaiseException: (TypeError) illegal access on 'process': class org.jruby.javasupport.JavaMethod cannot access a member of class org.asciidocto
r.extension.MacroProcessor with modifiers "protected abstract"                                                                                                                
    at org.jruby.RubyMethod.call (org/jruby/RubyMethod.java:127)                                                                                                              
    at RUBY.block in sub_macros (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/substitutors.rb:642)                                                               
    at org.jruby.RubyString.gsub (org/jruby/RubyString.java:2541)                                                                                                             
    at RUBY.block in sub_macros (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/substitutors.rb:615)                                                               
    at org.jruby.RubyArray.each (org/jruby/RubyArray.java:1734)                                                                                                               
    at RUBY.sub_macros (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/substitutors.rb:614)                                                                        
    at RUBY.block in apply_subs (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/substitutors.rb:126)                                                               
    at org.jruby.RubyArray.each (org/jruby/RubyArray.java:1734)                                                                                                               
    at RUBY.apply_subs (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/substitutors.rb:115)                                                                        
    at RUBY.content (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/block.rb:112)                                                                                  
    at RUBY.paragraph (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/converter/html5.rb:719)                                                                      
    at RUBY.convert (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/converter/base.rb:34)                                                                          
    at RUBY.convert (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/abstract_block.rb:70)                                                                          
    at RUBY.block in content (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/abstract_block.rb:79)                                                                 
    at org.jruby.RubyArray.map (org/jruby/RubyArray.java:2486)                                                                                                                
    at RUBY.content (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/abstract_block.rb:79)                                                                          
    at RUBY.content (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/document.rb:1120)                                                                              
    at RUBY.document (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/converter/html5.rb:186)                                                                       
    at RUBY.convert (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/converter/base.rb:34)                                                                          
    at RUBY.convert (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/document.rb:1060)                                                                              
    at RUBY.convert (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor.rb:1502)                                                                                       
    at RUBY.block in convert_file (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor.rb:1575)                                                                         
    at org.jruby.RubyIO.open (org/jruby/RubyIO.java:1156)                                                                                                                     
    at RUBY.convert_file (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor.rb:1575)                                                                                  
    at RUBY.convertFile (<script>:75)                                                                                                                                         
,,,

**Comment by jmini** [comment]
Does the change introduced with #625 not make all extensions incompatible with the new core of asciidoctorj?

At code level, I expect all macros to have following compile error:

> Cannot reduce the visibility of the inherited method from MacroProcessor

At bytecode level (having a macro with a protected method compiled against the old asciidoctor-core where the method is protected, and then executed with a new version of the asciidoctor-core), I do not know how it will behave.

I did not check it yet...

**Comment by robertpanzer** [comment]
Hmmm... That's a good point.
I probably have to revert the commit.
But that means that we will not be able to fix this problem easily.

(And at least imo it would be too much effort to introduce another wrapping layer with a public facade towards JRuby and keeping the old interface for the extension implementor, in particular as this problem doesn't seem to exist in 1.6.0)

**Comment by robertpanzer** [comment]
You're right about the error message. That was the reason I had to make the process method of the ManpageMacro in the tests public.

**Comment by jmini** [comment]
> I probably have to revert the commit.

Not necessary. This just needs to be properly communicated.
Finding an appropriate Version name will also be tricky.

**Comment by jmini** [comment]
Great news!

With the modifications (`process` method public in `MacroProcessor`), I could run AsciidoctorJ (version `1.5.7-SNAPSHOT` compiled locally) with an extension that has a `protected` method (compiled against an old version of Asciidoctor).

It works!
 (Tested with Java 8 and Java 9 and JRuby `9.1.15.0`).

---
In my opinion it is OK to let the commit as you have prepared it.

* For users using a compiled macro (dependency from a repository) it will work.
* For users compiling their dependencies, if they have the sources, fixing the “Cannot reduce the visibility“ compile error is not that complicated!

---
Thank you a lot for this change.

**Comment by abelsromero** [comment]
I am thinking if we should add java9 to Travis builds.

**Comment by robertpanzer** [comment]
> It works!
That's awesome news!
With this I think as well it's ok to keep the fix in the 1.5 branch.
Thanks for testing this.

Yes, we should try to get a Java9 build for master as well. We already have it for master.

**Comment by jmini** [comment]
Pull request to add java9 for the `1.5.7-SNAPSHOT` builds (`master` branch) => #628
